dll/win32/fmifs/query.c

```diff
@@ -43,3 +49,147 @@ QueryAvailableFileSystemFormat(
 
     return TRUE;
 }
+
+/**
+ * @brief
+ * Retrieves disk device information.
+ *
+ * @param[in] DriveRoot
+ * String which contains a DOS device name,
+ *
+ * @param[in,out] DeviceInformation
+ * Pointer to buffer with DEVICE_INFORMATION structure which will receive data.
+ *
+ * @param[in] BufferSize
+ * Size of buffer in bytes.
+ *
+ * @return
+ * TRUE if the buffer was large enough and was filled with
+ * the requested information, FALSE otherwise.
+ *
+ * @remarks
+ * The returned information is mostly related to Sony Memory Stick devices.
+ * On Vista+ the returned information is disk sector size and volume length in sectors,
+ * regardless of the type of disk.
+ * ReactOS implementation returns DEVICE_HOTPLUG flag if inspected device is a hotplug device
+ * as well as sector size and volume length of disk device.
+ */
+BOOL
+NTAPI
+QueryDeviceInformation(
+    _In_ PWCHAR DriveRoot,
+    _Out_ PVOID DeviceInformation,
+    _In_ ULONG BufferSize)
+{
+    PDEVICE_INFORMATION DeviceInfo = DeviceInformation;
+    IO_STATUS_BLOCK Iosb;
+    DISK_GEOMETRY DiskGeometry;
+    STORAGE_HOTPLUG_INFO HotplugInfo;
+    GET_LENGTH_INFORMATION LengthInformation;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    UNICODE_STRING DeviceName;
+    HANDLE FileHandle;
+    NTSTATUS Status;
+    WCHAR DiskDevice[64];
+    WCHAR DosPath[16];
+
+    /* Sanity checks, buffer should be able to at least hold DeviceFlags */
+    if (DriveRoot == NULL || BufferSize < sizeof(ULONG))
```

Comment:
If the first parameter DriveRoot is not optional, there's no need for a NULL check on it; the function has to crash (it's user-mode) if its caller misuses it.